Automattic/jetpack
# [Boost] Fix missing WP_Error class

Fixes pedxs5-oQ-p2#comment-309 .. plus more.

We got a report that some warnings could bubble up to the user's admin interface UI badly. What started as making sure that a `mkdir` call was consistently using `Filesystem_Utils` like the rest lead me down a rabbit-hole in testing.

It turns out that we were using `WP_Error` throughout our custom functions, which is not available during pre-WordPress. So I fixed that by introducing `Boost_Cache_Error` - a new class that fulfils a similar role. Why didn't I just load WP_Error early? Because the file it's in is `require`ed from wp-settings; not `require_once`ed - so it is not safe to load it early.

Once I got down that rabbit-hole, I discovered that trying to create a boost cache settings file every time we tried to read any setting lead us to spam errors creating a settings file if the dir is unwriteable. So, I've also made it only create a settings file when turning the cache feature on, or when writing a setting out.

Whew. It was a rabbit-hole. But all of this makes setup no longer barf errors when running on a host with an unwriteable wp-content/boost-cache directory.

## Proposed changes:
* Replace stray `mkdir` call with `Filesystem_Utils::create_directory`.
* Create the settings file when you turn on the Boost Cache feature, or when writing a setting if it doesn't already exist.
* Introduce a `Boost_Cache_Error` class we can use before `WP_Error` is loaded.
* Added in error_log to boost_cache_settings if it cannot init a settings file and WP_DEBUG is true. It cannot use its own logs in that situation, so the error had to be reported *somewhere*.
* Remove the last_error stuff from boost_cache_settings, as that was added as a workaround to not having a wp_error-like class to use.

### Other information:


## Jetpack product discussion


n/a

## Does this pull request change what data or activity we track or use?


no

## Testing instructions:





* Nuke your boost-cache directory, mess with boost cache and make sure it works as you'd expect
* Create an empty and unwriteable boost-cache directory and turn caching off and on. Ensure that a sensible error is shown, and no logs are barfed unreasonably.
* Generally mess around and make sure there aren't error_logs we didn't explicitly send, and that caching and its settings works.
